aten/src/ATen/native/cuda/Loops.cuh

```diff
@@ -131,6 +141,157 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
   return invoke_impl<traits>(f, data, strides, dtypes, i, Indices{});
 }
 
+// similar to above code but work on lambda using index for calculation
+template <typename traits, typename func_t, typename index_t, size_t... INDEX>
+C10_HOST_DEVICE typename traits::result_type
+invoke_with_index_impl(const func_t &f, char* const C10_RESTRICT data[], const index_t strides[], int i, int idx,
+                       std::index_sequence<INDEX...>) {
+  return f(*(typename traits::template arg<INDEX>::type*)(data[INDEX] + i * strides[INDEX])..., idx);
+}
+
+template <typename func_t, typename index_t, typename traits = function_traits<func_t>>
+C10_HOST_DEVICE typename traits::result_type
+invoke_with_index(const func_t &f, char* const C10_RESTRICT data[], const index_t strides[], int i, int idx) {
+  // index at last position
+  using Indices = std::make_index_sequence<traits::arity-1>;
+  return invoke_with_index_impl<traits>(f, data, strides, i, idx, Indices{});
+}
+
+template <typename traits, typename func_t, typename index_t, size_t... I>
+C10_HOST_DEVICE typename traits::result_type
+invoke_with_index_impl(const func_t &f, char* const C10_RESTRICT data[], const index_t strides[], const ScalarType dtypes[],
+                       int i, int idx, std::index_sequence<I...>) {
+  return f(c10::fetch_and_cast<typename traits::template arg<I>::type>(dtypes[I], data[I] + i * strides[I])..., idx);
+}
+
+template <typename func_t, typename index_t, typename traits = function_traits<func_t>>
+C10_HOST_DEVICE typename traits::result_type
+invoke_with_index(const func_t &f, char* const C10_RESTRICT data[], const index_t strides[], const ScalarType dtypes[],
+                  int i, int idx) {
+  // index at last position
+  using Indices = std::make_index_sequence<traits::arity-1>;
+  return invoke_with_index_impl<traits>(f, data, strides, dtypes, i, idx, Indices{});
+}
+
+} // namespace legacy
+
+// See the note for namespace legacy above.
+namespace modern {
+
+namespace detail {
+
+template <typename func_t, typename array_t, std::size_t... I>
+__device__ inline constexpr decltype(auto) apply_impl(func_t f, array_t t, std::index_sequence<I...>)
+{
+    return f(t[I]...);
+}
+template <typename func_t, typename array_t>
+__device__ inline constexpr decltype(auto) array_apply(func_t f, array_t a) {
```

Comment:
nit: would it make sense to call these functions `invoke_with_array` and `invoke_with_array_impl`, to follow the others?

Reply:
Yes, that name sounds better. But let me rename it in the next PR.

Reply:
This has merge conflicts... So I will fix it here then.